src/core/ddsc/src/dds_whc_builtintopic.c

```diff
@@ -86,49 +90,107 @@ static bool bwhc_sample_iter_borrow_next (struct whc_sample_iter *opaque_it, str
+        case DSBT_TOPIC:       kind = EK_TOPIC; break;
         case DSBT_WRITER:      kind = EK_WRITER; break;
         case DSBT_READER:      kind = EK_READER; break;
       }
-      assert (whc->type == DSBT_PARTICIPANT || kind != EK_PARTICIPANT);
+      assert (whc->entity_kind == DSBT_PARTICIPANT || kind != EK_PARTICIPANT);
       entidx_enum_init (&it->it, whc->entidx, kind);
       it->st = BIS_LOCAL;
       /* FALLS THROUGH */
     case BIS_LOCAL:
       while ((entity = entidx_enum_next (&it->it)) != NULL)
         if (is_visible (entity))
           break;
-      if (entity) {
-        sample->serdata = dds__builtin_make_sample (entity, entity->tupdate, true);
+      if (entity)
+      {
+#ifdef DDSI_INCLUDE_TOPIC_DISCOVERY
+        if (kind == EK_TOPIC)
+        {
+          sample->serdata = dds__builtin_make_sample_topic (((struct topic *)entity)->definition, entity->tupdate, true);
+        }
+        else
+#endif
+        {
+          sample->serdata = dds__builtin_make_sample (entity, entity->tupdate, true);
+        }
         it->have_sample = true;
         return true;
-      } else {
-        entidx_enum_fini (&it->it);
-        it->st = BIS_INIT_PROXY;
       }
+      entidx_enum_fini (&it->it);
+      it->st = BIS_INIT_PROXY;
       /* FALLS THROUGH */
     case BIS_INIT_PROXY:
-      switch (whc->type) {
-        case DSBT_PARTICIPANT: kind = EK_PROXY_PARTICIPANT; break;
-        case DSBT_WRITER:      kind = EK_PROXY_WRITER; break;
-        case DSBT_READER:      kind = EK_PROXY_READER; break;
+#ifdef DDSI_INCLUDE_TOPIC_DISCOVERY
+      if (whc->entity_kind == DSBT_TOPIC)
+      {
+        /* proxy topics are not stored in entity index as these are not real
+           entities. For proxy topics loop over all proxy participants and
+           iterate all proxy topics for each proxy participant*/
+        entidx_enum_init (&it->it, whc->entidx, EK_PROXY_PARTICIPANT);
+        it->cur_proxypp = NULL;
       }
-      assert (kind != EK_PARTICIPANT);
-      entidx_enum_init (&it->it, whc->entidx, kind);
+      else
+#endif
+      {
+        switch (whc->entity_kind)
+        {
+          case DSBT_PARTICIPANT: kind = EK_PROXY_PARTICIPANT; break;
+          case DSBT_TOPIC:       assert (0); break;
+          case DSBT_WRITER:      kind = EK_PROXY_WRITER; break;
+          case DSBT_READER:      kind = EK_PROXY_READER; break;
+        }
+        assert (kind != EK_PARTICIPANT);
+        entidx_enum_init (&it->it, whc->entidx, kind);
+      }
+
       it->st = BIS_PROXY;
       /* FALLS THROUGH */
     case BIS_PROXY:
-      while ((entity = entidx_enum_next (&it->it)) != NULL)
-        if (is_visible (entity))
-          break;
-      if (entity) {
+#ifdef DDSI_INCLUDE_TOPIC_DISCOVERY
+      if (whc->entity_kind == DSBT_TOPIC)
+      {
+        struct proxy_topic *proxytp = NULL;
+        if (it->cur_proxypp != NULL)
+        {
+          ddsrt_mutex_lock (&it->cur_proxypp->e.lock);
```

Comment:
There's a little smell here. You take the lock conditionally, but unlock unconditionally. I think it would be cleaner to move this into the `while` loop to?

Reply:
There was an unnecessary `if` and an unreachable `return false` after the `while` loop, I've removed that in bf8af34. I think the conditions for lock and unlock are correct: the first lock, and unlock in case there is no next topic for the current proxypp, should be conditional because this should not be done in the first iteration. The last unlock can only be reached with `it->cur_proxypp->e.lock` taken, so there is no condition needed there. I've added some code comments in bf8af34 that hopefully make the code a bit more clear.